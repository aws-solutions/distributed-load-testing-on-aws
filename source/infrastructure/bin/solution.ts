// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export const SOLUTIONS_METRICS_ENDPOINT = "https://metrics.awssolutionsbuilder.com/generic";

export class Solution {
  public readonly id: string;
  public readonly name: string;
  public readonly version: string;
  public description: string;

  constructor(id: string, name: string, version: string, description: string) {
    this.id = id;
    this.name = name;
    this.version = version;
    this.description = description;
  }
}
