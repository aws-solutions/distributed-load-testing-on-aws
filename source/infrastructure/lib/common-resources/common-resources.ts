// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnResource, Fn, RemovalPolicy, Stack, Tags, CfnCondition } from "aws-cdk-lib";
import {
  BlockPublicAccess,
  Bucket,
  BucketAccessControl,
  BucketEncryption,
  IBucket,
  ObjectOwnership,
} from "aws-cdk-lib/aws-s3";
import { AnyPrincipal, Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import * as appreg from "@aws-cdk/aws-servicecatalogappregistry-alpha";
import { Construct } from "constructs";
import {
  CfnResourceAssociation,
  CfnApplication,
  CfnAttributeGroup,
  CfnAttributeGroupAssociation,
} from "aws-cdk-lib/aws-servicecatalogappregistry";

export interface CommonResourcesConstructProps {
  readonly sourceCodeBucket: string;
}

export interface AppRegistryApplicationProps {
  readonly description: string;
  readonly stackType: string;
  readonly solutionId: string;
  readonly applicationName: string;
  readonly solutionVersion: string;
}

/**
 * Distributed Load Testing on AWS common resources construct.
 * Creates a CloudWatch logs policy and an S3 bucket to store logs.
 */
export class CommonResourcesConstruct extends Construct {
  // CloudWatch logs Policy
  public cloudWatchLogsPolicy: Policy;
  // Code S3 Bucket
  public sourceBucket: IBucket;

  constructor(scope: Construct, id: string, props: CommonResourcesConstructProps) {
    super(scope, id);

    const logGroupResourceArn = Stack.of(this).formatArn({
      service: "logs",
      resource: "log-group:",
      resourceName: "aws/lambda/*",
    });
    this.cloudWatchLogsPolicy = new Policy(this, "CloudWatchLogsPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["logs:CreateLogGroup", "logs:CreateLogStream", "logs:PutLogEvents"],
          resources: [logGroupResourceArn],
        }),
      ],
    });

    this.sourceBucket = Bucket.fromBucketName(this, "SourceCodeBucket", props.sourceCodeBucket);
  }

  public s3LogsBucket(): Bucket {
    const logsBucket = new Bucket(this, "LogsBucket", {
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN,
      enforceSSL: true,
      versioned: true,
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
    });

    logsBucket.addToResourcePolicy(
      new PolicyStatement({
        actions: ["s3:*"],
        conditions: {
          Bool: { "aws:SecureTransport": "false" },
        },
        effect: Effect.DENY,
        principals: [new AnyPrincipal()],
        resources: [logsBucket.bucketArn, logsBucket.arnForObjects("*")],
      })
    );

    const s3LogsBucketResource = logsBucket.node.defaultChild as CfnResource;
    s3LogsBucketResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W35",
          reason: "This is the logging bucket, it does not require logging.",
        },
        {
          id: "W51",
          reason: "Since the bucket does not allow the public access, it does not require to have bucket policy.",
        },
      ],
    });
    return logsBucket;
  }

  public appRegistryApplication(props: AppRegistryApplicationProps) {
    const stack = Stack.of(this);
    const applicationType = "AWS-Solutions";
    const solutionName = "Distributed Load Testing";

    const application = new appreg.Application(stack, "AppRegistry", {
      applicationName: Fn.join("-", [props.applicationName, Aws.REGION, Aws.ACCOUNT_ID]),
      description: `Service Catalog application to track and manage all your resources for the solution ${solutionName}`,
    });
    application.associateApplicationWithStack(stack);

    Tags.of(application).add("Solutions:SolutionID", props.solutionId);
    Tags.of(application).add("Solutions:SolutionName", solutionName);
    Tags.of(application).add("Solutions:SolutionVersion", props.solutionVersion);
    Tags.of(application).add("Solutions:ApplicationType", applicationType);

    const attributeGroup = new appreg.AttributeGroup(stack, "DefaultApplicationAttributes", {
      attributeGroupName: Aws.STACK_NAME,
      description: "Attribute group for solution information",
      attributes: {
        applicationType,
        version: props.solutionVersion,
        solutionID: props.solutionId,
        solutionName,
      },
    });
    attributeGroup.associateWith(application);
    // Add hard-coded conditions to the AppRegistry resources
    // These can be later on changed by users if they failed to update the
    // Stack due to change of AppRegistry resources logical Id
    this.addConditions(stack, [
      CfnApplication,
      CfnAttributeGroup,
      CfnAttributeGroupAssociation,
      CfnResourceAssociation,
    ]);
  }

  private addConditions(stack: Stack, resourceTypes: Array<any>) {
    const createAppCondition = new CfnCondition(this, "AppRegistryCondition", {
      expression: Fn.conditionEquals("true", "true"),
    });

    const visit = (node: Construct) => {
      for (const resourceType of resourceTypes) {
        if (node instanceof resourceType) {
          (node as CfnResource).cfnOptions.condition = createAppCondition;
        }
      }
      for (const child of node.node.children) {
        visit(child as Construct);
      }
    };
    visit(stack);
  }
}
