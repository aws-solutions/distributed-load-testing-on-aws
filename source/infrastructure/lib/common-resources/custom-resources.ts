// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnCondition, CfnCustomResource, CustomResource } from "aws-cdk-lib";
import { Construct } from "constructs";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { IBucket } from "aws-cdk-lib/aws-s3";

export interface ConsoleConfigProps {
  readonly apiEndpoint: string;
  readonly cognitoIdentityPool: string;
  readonly cognitoUserPool: string;
  readonly cognitoUserPoolClient: string;
  readonly consoleBucket: IBucket;
  readonly scenariosBucket: string;
  readonly iotEndpoint: string;
  readonly iotPolicy: string;
}

export interface SendAnonymizedMetricsCRProps {
  readonly existingVpc: string;
  readonly sendAnonymizedUsage: string;
  readonly sendAnonymizedUsageCondition: CfnCondition;
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly uuid: string;
}

export interface TestingResourcesConfigCRProps {
  readonly taskCluster: string;
  readonly ecsCloudWatchLogGroup: string;
  readonly taskSecurityGroup: string;
  readonly taskDefinition: string;
  readonly subnetA: string;
  readonly subnetB: string;
}

export interface PutRegionalTemplateProps {
  readonly sourceCodeBucketName: string;
  readonly regionalTemplatePrefix: string;
  readonly mainStackRegion: string;
  readonly scenariosTable: string;
  readonly lambdaTaskRoleArn: string;
  readonly scenariosBucket: string;
}
export interface DetachIotPrincipalPolicyProps {
  readonly iotPolicyName: string;
}

/**
 * Distributed Load Testing on AWS Custom Resources Construct.
 * It creates a custom resource Lambda function, a solution UUID, and a custom resource to send anonymized usage.
 */
export class CustomResourcesConstruct extends Construct {
  private readonly customResourceLambda: NodejsFunction;

  constructor(scope: Construct, id: string, customResourceLambda: NodejsFunction) {
    super(scope, id);
    this.customResourceLambda = customResourceLambda;
  }

  private createCustomResource(
    id: string,
    customResourceFunction: NodejsFunction,
    props?: Record<string, unknown>
  ): CustomResource {
    return new CustomResource(this, id, {
      serviceToken: customResourceFunction.functionArn,
      properties: props,
    });
  }

  public getIotEndpoint() {
    const iotEndpoint = this.createCustomResource("GetIotEndpoint", this.customResourceLambda, {
      Resource: "GetIotEndpoint",
    });
    return iotEndpoint.getAtt("IOT_ENDPOINT").toString();
  }

  public detachIotPrincipalPolicy(props: DetachIotPrincipalPolicyProps) {
    this.createCustomResource("DetachIotPrincipalPolicy", this.customResourceLambda, {
      Resource: "DetachIotPolicy",
      IotPolicyName: props.iotPolicyName,
    });
  }

  public putRegionalTemplate(props: PutRegionalTemplateProps) {
    this.createCustomResource("PutRegionalTemplate", this.customResourceLambda, {
      Resource: "PutRegionalTemplate",
      SrcBucket: props.sourceCodeBucketName,
      SrcPath: props.regionalTemplatePrefix,
      DestBucket: props.scenariosBucket,
      MainRegionLambdaTaskRoleArn: props.lambdaTaskRoleArn,
      ScenariosTable: props.scenariosTable,
      MainRegionStack: props.mainStackRegion,
    });
  }

  public consoleConfig(props: ConsoleConfigProps) {
    const awsExports = `const awsConfig = {
      aws_iot_endpoint: '${props.iotEndpoint}',
      aws_iot_policy_name: '${props.iotPolicy}',
      cw_dashboard: 'https://console.aws.amazon.com/cloudwatch/home?region=${Aws.REGION}#dashboards:',
      ecs_dashboard: 'https://${Aws.REGION}.console.aws.amazon.com/ecs/home?region=${Aws.REGION}#/clusters/${Aws.STACK_NAME}/tasks',
      aws_project_region: '${Aws.REGION}',
      aws_cognito_region: '${Aws.REGION}',
      aws_cognito_identity_pool_id: '${props.cognitoIdentityPool}',
      aws_user_pools_id: '${props.cognitoUserPool}',
      aws_user_pools_web_client_id: '${props.cognitoUserPoolClient}',
      oauth: {},
      aws_cloud_logic_custom: [
          {
              name: 'dlts',
              endpoint: '${props.apiEndpoint}',
              region: '${Aws.REGION}'
          }
      ],
      aws_user_files_s3_bucket: '${props.scenariosBucket}',
      aws_user_files_s3_bucket_region: '${Aws.REGION}',
  }`;
    this.customResourceLambda.role?.addToPrincipalPolicy(
      new PolicyStatement({
        actions: ["s3:PutObject"],
        resources: [`${props.consoleBucket.bucketArn}/*`],
        effect: Effect.ALLOW,
      })
    ); // need put permission to upload config with deployment specific values
    this.createCustomResource("ConsoleConfig", this.customResourceLambda, {
      AwsExports: awsExports,
      DestBucket: props.consoleBucket.bucketName,
      Resource: "ConfigFile",
    });
  }

  public uuidGenerator() {
    const uuid = this.createCustomResource("CustomResourceUuid", this.customResourceLambda, {
      Resource: "UUID",
    });
    return {
      uuid: uuid.getAttString("UUID").toString(),
      suffix: uuid.getAttString("SUFFIX").toString(),
    };
  }

  public sendAnonymizedMetricsCR(props: SendAnonymizedMetricsCRProps) {
    const sendAnonymizedMetrics = this.createCustomResource("AnonymizedMetric", this.customResourceLambda, {
      existingVPC: props.existingVpc,
      Region: Aws.REGION,
      Resource: "AnonymizedMetric",
      SolutionId: props.solutionId,
      UUID: props.uuid,
      Version: props.solutionVersion,
    });
    const cfnSendAnonymizedMetrics = sendAnonymizedMetrics.node.defaultChild as CfnCustomResource;
    cfnSendAnonymizedMetrics.cfnOptions.condition = props.sendAnonymizedUsageCondition;
  }

  public testingResourcesConfigCR(props: TestingResourcesConfigCRProps) {
    const testingResourcesConfig = {
      region: Aws.REGION,
      subnetA: props.subnetA,
      subnetB: props.subnetB,
      ecsCloudWatchLogGroup: props.ecsCloudWatchLogGroup,
      taskSecurityGroup: props.taskSecurityGroup,
      taskDefinition: props.taskDefinition,
      taskImage: `${Aws.STACK_NAME}-load-tester`,
      taskCluster: props.taskCluster,
    };
    this.createCustomResource("TestingResourcesConfig", this.customResourceLambda, {
      TestingResourcesConfig: testingResourcesConfig,
      Resource: "TestingResourcesConfigFile",
    });
  }
}
