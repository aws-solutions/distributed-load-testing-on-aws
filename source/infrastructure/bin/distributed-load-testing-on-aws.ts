// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { App, DefaultStackSynthesizer } from "aws-cdk-lib";
import { DLTStack } from "../lib/distributed-load-testing-on-aws-stack";
import { Solution } from "./solution";
import { RegionalInfrastructureDLTStack } from "../lib/distributed-load-testing-on-aws-regional-stack";

// CDK and default deployment
let synthesizer = new DefaultStackSynthesizer({ generateBootstrapVersionRule: false });

// Solutions pipeline deployment
const { DIST_OUTPUT_BUCKET, SOLUTION_ID, SOLUTION_NAME, VERSION, PUBLIC_ECR_REGISTRY, PUBLIC_ECR_TAG } = process.env;
if (DIST_OUTPUT_BUCKET && SOLUTION_NAME && VERSION && PUBLIC_ECR_REGISTRY && PUBLIC_ECR_TAG)
  synthesizer = new DefaultStackSynthesizer({
    generateBootstrapVersionRule: false,
    fileAssetsBucketName: `${DIST_OUTPUT_BUCKET}-\${AWS::Region}`,
    bucketPrefix: `${SOLUTION_NAME}/${VERSION}/`,
    imageAssetsRepositoryName: PUBLIC_ECR_REGISTRY,
    dockerTagPrefix: PUBLIC_ECR_TAG,
  });

const app = new App();
const solutionName = app.node.tryGetContext("solutionName");
const solutionVersion = VERSION ?? app.node.tryGetContext("solutionVersion");
const solutionId = SOLUTION_ID ?? app.node.tryGetContext("solutionId");
const mainStackDescription = `(${solutionId}) - ${solutionName}. Version ${solutionVersion}`;
const regionsStackDescription = `(${solutionId}-regional) - Distributed Load Testing on AWS testing resources regional deployment. Version ${solutionVersion}`;

const solution = new Solution(solutionId, solutionName, solutionVersion, mainStackDescription);

// main stack
new DLTStack(app, "distributed-load-testing-on-aws", {
  synthesizer,
  solution,
  stackType: "main",
});

// regional stack
solution.description = regionsStackDescription;
new RegionalInfrastructureDLTStack(app, "distributed-load-testing-on-aws-regional", {
  synthesizer,
  solution,
  stackType: "regional",
});
