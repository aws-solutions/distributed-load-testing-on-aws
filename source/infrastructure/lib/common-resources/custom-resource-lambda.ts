// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Construct } from "constructs";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Effect, Policy, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { ArnFormat, CfnResource, Duration, Stack } from "aws-cdk-lib";
import { Solution, SOLUTIONS_METRICS_ENDPOINT } from "../../bin/solution";
import * as path from "path";
import { addCfnGuardSuppression } from "./add-cfn-guard-suppression";

interface ICustomResourceLambda {
  nodejsLambda: NodejsFunction;
  addEnvironmentVariables(environmentVariables: { [key: string]: string }): void;
  addPolicy(policy: Policy[]): void;
}

export class CustomResourceLambda extends Construct implements ICustomResourceLambda {
  public readonly nodejsLambda: NodejsFunction;

  constructor(scope: Construct, id: string, solution: Solution) {
    super(scope, id);

    const policyForCustomResource = new PolicyDocument({
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["iot:DescribeEndpoint", "iot:DetachPrincipalPolicy"],
          resources: ["*"],
        }),
        new PolicyStatement({
          actions: ["iot:ListTargetsForPolicy"],
          effect: Effect.ALLOW,
          resources: [
            Stack.of(this).formatArn({
              service: "iot",
              resource: "policy",
              resourceName: "*",
              arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
            }),
          ],
        }),
      ],
    });

    const customResourceRole = new Role(this, "CustomResourceLambdaRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        CustomResourcePolicy: policyForCustomResource,
      },
    });

    const cfnCustomResourceRole = customResourceRole.node.defaultChild as CfnResource;
    cfnCustomResourceRole.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "iot:DescribeEndpoint and iot:DetachPrincipalPolicy cannot specify the resource.",
        },
      ],
    });

    addCfnGuardSuppression(customResourceRole, "IAM_NO_INLINE_POLICY_CHECK");

    this.nodejsLambda = new NodejsFunction(this, "CustomResourceLambda", {
      description: "CFN Lambda backed custom resource to deploy assets to s3",
      handler: "index.handler",
      role: customResourceRole,
      entry: path.join(__dirname, "../../../custom-resource/main-index.js"),
      runtime: Runtime.NODEJS_20_X,
      timeout: Duration.seconds(120),
      environment: {
        METRIC_URL: SOLUTIONS_METRICS_ENDPOINT,
        SOLUTION_ID: solution.id,
        VERSION: solution.version,
      },
    });

    (this.nodejsLambda.node.defaultChild as CfnResource).overrideLogicalId(
      "DLTCustomResourceInfraCustomResourceLambdaA4053269"
    );
    (this.nodejsLambda.node.defaultChild as CfnResource).addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "VPC not needed for lambda",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });
  }

  public addEnvironmentVariables(environmentVariables: { [key: string]: string }) {
    for (const [key, value] of Object.entries(environmentVariables)) {
      this.nodejsLambda.addEnvironment(key, value);
    }
  }

  public addPolicy(policy: Policy[]) {
    for (const _policy of policy) {
      this.nodejsLambda.role?.attachInlinePolicy(_policy);
    }
  }
}
