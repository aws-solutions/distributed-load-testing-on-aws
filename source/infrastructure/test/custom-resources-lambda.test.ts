// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Template } from "aws-cdk-lib/assertions";
import { App, DefaultStackSynthesizer, Stack } from "aws-cdk-lib";
import { CustomResourceLambda } from "../lib/common-resources/custom-resource-lambda";
import { Solution, SOLUTIONS_METRICS_ENDPOINT } from "../bin/solution";
import { createTemplateWithoutS3Key } from "./snapshot_helpers";

test("DLT API Test", () => {
  const app = new App();
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
    }),
  });

  const solution = new Solution("testId", "DLT", "testVersion", "mainStackDescription");
  new CustomResourceLambda(stack, "TestCustomResourceInfra", solution);

  expect(createTemplateWithoutS3Key(stack)).toMatchSnapshot();
  Template.fromStack(stack).hasResourceProperties("AWS::Lambda::Function", {
    Description: "CFN Lambda backed custom resource to deploy assets to s3",
    Environment: {
      Variables: {
        METRIC_URL: SOLUTIONS_METRICS_ENDPOINT,
        SOLUTION_ID: solution.id,
        VERSION: solution.version,
      },
    },
    Handler: "index.handler",
    Runtime: "nodejs20.x",
    Timeout: 120,
  });
});
