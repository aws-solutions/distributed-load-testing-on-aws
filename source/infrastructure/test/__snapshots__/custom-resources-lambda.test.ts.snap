// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT API Test 1`] = `
{
  "Resources": {
    "DLTCustomResourceInfraCustomResourceLambdaA4053269": {
      "DependsOn": [
        "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            {
              "id": "W89",
              "reason": "VPC not needed for lambda",
            },
            {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "Omitted to remove snapshot dependency on hash",
        },
        "Description": "CFN Lambda backed custom resource to deploy assets to s3",
        "Environment": {
          "Variables": {
            "METRIC_URL": "https://metrics.awssolutionsbuilder.com/generic",
            "SOLUTION_ID": "testId",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestCustomResourceInfraCustomResourceLambdaRole03671AE8": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "iot:DescribeEndpoint and iot:DetachPrincipalPolicy cannot specify the resource.",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "iot:DescribeEndpoint",
                    "iot:DetachPrincipalPolicy",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "iot:ListTargetsForPolicy",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":policy/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CustomResourcePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
